src/System.IO.Pipelines/tests/PipeReaderWriterFacts.cs

```diff
@@ -775,6 +776,44 @@ public void GetMemoryZeroReturnsNonEmpty()
             Assert.True(_pipe.Writer.GetMemory(0).Length > 0);
         }
 
+        [Fact]
+        public async Task ReadAsyncWithDataReadyReturnsTaskWithValue()
+        {
+            _pipe.Writer.WriteEmpty(10);
+            await _pipe.Writer.FlushAsync();
+            var task = _pipe.Reader.ReadAsync();
+            Assert.True(IsTaskWithResult(task));
+        }
+
+        [Fact]
+        public void CancelledReadAsyncReturnsTaskWithValue()
+        {
+            _pipe.Reader.CancelPendingRead();
+            var task = _pipe.Reader.ReadAsync();
+            Assert.True(IsTaskWithResult(task));
+        }
+
+        [Fact]
+        public void FlushAsyncWithoutBackpressureReturnsTaskWithValue()
+        {
+            _pipe.Writer.WriteEmpty(10);
+            var task = _pipe.Writer.FlushAsync();
+            Assert.True(IsTaskWithResult(task));
+        }
+
+        [Fact]
+        public void CancelledFlushAsyncReturnsTaskWithValue()
+        {
+            _pipe.Writer.CancelPendingFlush();
+            var task = _pipe.Writer.FlushAsync();
+            Assert.True(IsTaskWithResult(task));
+        }
+
+        private bool IsTaskWithResult<T>(ValueTask<T> task)
+        {
+            return task.GetType().GetField("_result", BindingFlags.Instance | BindingFlags.NonPublic).GetValue(task) != null;
```

Comment:
@stephentoub is there a better way to check that ValueTask contains TResult?

Reply:
Ugh... This is way too hacky. We should verify if anything else that implements ValueTask even tests this. If not, it's not worth it. That same argument could be made about ValueTask\<T\> backed by a Task vs T

Reply:
I think it's important thing to verify but agree that this way is a bit too extreme.

Reply:
I think we should get rid of these tests.

Reply:
> is there a better way to check that ValueTask contains TResult?

They way you're using it, you can do:
```C#
private bool IsTaskWithResult<T>(ValueTask<T> task) => task == new ValueTask<T>(task.Result);
```